piptools/utils.py

```diff
@@ -522,3 +531,126 @@ def parse_requirements_from_wheel_metadata(
+    else:
+        config_file = Path(value)
+
+    try:
+        config = parse_config_file(config_file)
+    except OSError as e:
+        raise click.FileError(
+            filename=str(config_file), hint=f"Could not read '{config_file}': {e}"
+        )
+    except ValueError as e:
+        raise click.FileError(
+            filename=str(config_file), hint=f"Could not parse '{config_file}': {e}"
+        )
+
+    if not config:
+        return None
+
+    defaults: dict[str, Any] = ctx.default_map.copy() if ctx.default_map else {}
+    defaults.update(config)
+
+    ctx.default_map = defaults
+    return str(config_file)
+
+
+def select_config_file(src_files: tuple[str, ...]) -> Path | None:
+    """
+    Returns the config file to use for defaults given `src_files` provided.
+    """
+    if not src_files:
+        # If no src_files were specified, we consider the current directory the only candidate
+        candidate_dirs = [Path.cwd()]
+    else:
+        # Collect the candidate directories based on the src_file arguments provided
+        src_files_as_paths = [
+            Path(Path.cwd(), src_file).resolve() for src_file in src_files
+        ]
+        candidate_dirs = [
+            src if src.is_dir() else src.parent for src in src_files_as_paths
+        ]
+    config_file_path = next(
+        (
+            candidate_dir / config_file
+            for candidate_dir in candidate_dirs
+            for config_file in (CONFIG_FILE_NAME, "pyproject.toml")
+            if (candidate_dir / config_file).is_file()
+        ),
+        None,
+    )
+    return config_file_path
+
+
+# Some of the defined click options have different `dest` values than the defaults
+NON_STANDARD_OPTION_DEST_MAP: dict[str, str] = {
+    "extra": "extras",
+    "upgrade_package": "upgrade_packages",
+    "resolver": "resolver_name",
+    "user": "user_only",
+}
+
+
+def get_click_dest_for_option(option_name: str) -> str:
+    """Returns the click `dest` value for the given option name."""
+    # Format the keys properly
+    option_name = option_name.lstrip("-").replace("-", "_").lower()
+    # Some options have dest values that are overrides from the click generated default
+    option_name = NON_STANDARD_OPTION_DEST_MAP.get(option_name, option_name)
+    return option_name
+
+
+# Ensure that any default overrides for these click options are lists, supporting multiple values
+MULTIPLE_VALUE_OPTIONS = [
+    "extras",
+    "upgrade_packages",
+    "unsafe_package",
+    "find_links",
+    "extra_index_url",
+    "trusted_host",
+]
+
+
+@functools.lru_cache()
+def parse_config_file(config_file: Path) -> dict[str, Any]:
+    if sys.version_info >= (3, 11):
+        # Python 3.11 stdlib tomllib load() requires a binary file object
+        with config_file.open("rb") as ifs:
+            config = tomllib.load(ifs)
+    else:
+        # Before 3.11, using the external toml library, load requires the filename
+        config = toml.load(str(config_file))
```

Comment:
Having seen this, I remembered that CPython specifically chose to mimic the popular library APIs. I also recall that all the major ecosystem projects moving away from some toml libraries. Namely, `toml` was something that stopped being used, along with `tomlkit`. I think, I still have one project using `toml` that I have to migrate.
I've verified that the library you should be using is called `tomli`. I checked that it's now used by `pip`, `pypa/build`, `pytest` and `tox`, among others.

Once you fix that, it'll be possible to make this somewhat more elegant:
```suggestion
    config = tomllib.loads(config_file.read_text(encoding="utf-8"))
```